### ISSUE TITLE
How do I convert a curve point into a public key?

### ISSUE THREAD

**Issue by AljoschaMeyer** [paulmillr/noble-curves#137]
If I have an x25519 curve point (say, returned from `hashToCurve` from `"npm:@noble/curves/ed25519"`), how do I convert it into a corresponding ed25519 public key? Does `curvePoint.toRawBytes()` yield the public key, or something else? I cannot find any documentation on what exactly `toRawBytes` actually returns.

**Comment by paulmillr** [comment]
1. x25519 (curve25519) is a different curve from ed25519. they are used for different cases
2. “Does curvePoint.toRawBytes() yield the public key, or something else?” - nope. It simply returns the encoded point.
3. hashToCurve does not really “encode to public key”. it encodes to point

can you clarify what’s your use case?

**Comment by AljoschaMeyer** [comment]
Thank you both for your replies 💜 

@nflatrea Yes, if this is not implemented by noble, we'll just have to implement that part of the ed25519 spec ourselves. I was simply hoping to get around that.

@paulmillr Good point asking for clarification, We *might* not actually need this.

TLDR: We need to *somehow* transmit curve points over the wire, and converting them into public keys seems like a well-specified and widely-implemented choice.

There are two independent use-cases: We want to use curve25519 for a [multiset-homomorphic hash function](https://willowprotocol.org/specs/3d-range-based-set-reconciliation/index.html#d3rbsr_parameters), and we want to use curve25519 to implement [private equality testing](https://willowprotocol.org/specs/pai/index.html#private_equality_testing).

For the homomorphic hash function, the underlying idea is simple: to hash a set of items, hash every individual item into the curve (via [curve25519XMD:SHA-512_ELL2_RO](https://www.rfc-editor.org/rfc/rfc9380#name-suites-for-curve25519-and-e), i.e., `hashToCurve`), then add the curve points together to obtain the hash of the set. But then, as a final step, we need to transmit the result over the wire. To do so, we need to encode the curve point into a sequence of bytes. My impression was that converting it into a public key would be a convenient choice of encoding, since that conversion would be part of the curve25519 spec, and hence widely implemented.

In principle, we could also just use noble's `toRawBytes`, but we would need to specify the exact behavior, not just refer to a javascript function that does not precisely specify what it does - this is for writing a [specification (draft state)](https://macromania--es6.deno.dev/earthstar/spec/index.html#es6_wgps_reconciliation), not just for hacking on a codebase.

For the privat equality testing part, two peers need to exchange curve points and perform scalar multiplication on them, which again requires encoding them over the wire. Since crypto libraries often implement scalar multiplication functions that take an encoded public key as input (or at least, some raw byte array and not a curve point struct, compare, e.g., [libsodium scalar_mult](https://libsodium.gitbook.io/doc/advanced/scalar_multiplication)), it again seemed sensible to simply exchange public keys over the wire. But before we can do that, we must first hash the items for which we want to test for set intersection into the curve. Which is again a job for RFC9380, but leaves us with a curve point that we now need to encode into a public key to send to the other peer.

Here again, we don't *necessarily* need to use public keys. We could also specify our own encoding of curve points, send that over the wire, and the other peer could decode and do scalar multiplication (with APIs that operate on actual curve points and scalars, not keys).

**Comment by paulmillr** [comment]
well, all ed25519 pubkeys are encoded in a way that is specified in toRawBytes. so it is a “public key”. it’s just that a concept of public key is different from a concept of point. public key has a specific private key

but again you need curve25519, not ed25519
